sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/OrderedListUserState.java

```diff
@@ -0,0 +1,314 @@
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.Range;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.TreeRangeSet;
+import org.joda.time.Instant;
+
+/**
+ * An implementation of an ordered list user state that utilizes the Beam Fn State API to fetch,
+ * clear and persist values.
+ *
+ * <p>Calling {@link #asyncClose()} schedules any required persistence changes. This object should
+ * no longer be used after it is closed.
+ *
+ * <p>TODO: Move to an async persist model where persistence is signalled based upon cache memory
+ * pressure and its need to flush.
+ */
+public class OrderedListUserState<T> {
+  private final BeamFnStateClient beamFnStateClient;
+  private final StateRequest request;
+  private final TimestampedValueCoder<T> timestampedValueCoder;
+  // Pending updates to persistent storage
+  private NavigableMap<Instant, Collection<T>> pendingAdds = Maps.newTreeMap();
+  private TreeRangeSet<Instant> pendingRemoves = TreeRangeSet.create();
+
+  private boolean isCleared = false;
+  private boolean isClosed = false;
+
+  public static class TimestampedValueCoder<T> extends StructuredCoder<TimestampedValue<T>> {
+
+    private final Coder<T> valueCoder;
+
+    // Internally, a TimestampedValue is encoded with a KvCoder, where the key is encoded with
+    // a VarLongCoder and the value is encoded with a LengthPrefixCoder.
+    // Refer to the comment in StateAppendRequest
+    // (org/apache/beam/model/fn_execution/v1/beam_fn_api.proto) for more detail.
+    private final KvCoder<Long, T> internalKvCoder;
+
+    public static <T> OrderedListUserState.TimestampedValueCoder<T> of(Coder<T> valueCoder) {
+      return new OrderedListUserState.TimestampedValueCoder<>(valueCoder);
+    }
+
+    @Override
+    public Object structuralValue(TimestampedValue<T> value) {
+      Object structuralValue = valueCoder.structuralValue(value.getValue());
+      return TimestampedValue.of(structuralValue, value.getTimestamp());
+    }
+
+    @SuppressWarnings("unchecked")
+    TimestampedValueCoder(Coder<T> valueCoder) {
+      this.valueCoder = checkNotNull(valueCoder);
+      this.internalKvCoder = KvCoder.of(VarLongCoder.of(), LengthPrefixCoder.of(valueCoder));
+    }
+
+    @Override
+    public void encode(TimestampedValue<T> windowedElem, OutputStream outStream)
+        throws IOException {
+      internalKvCoder.encode(
+          KV.of(windowedElem.getTimestamp().getMillis(), windowedElem.getValue()), outStream);
+    }
+
+    @Override
+    public TimestampedValue<T> decode(InputStream inStream) throws IOException {
+      KV<Long, T> kv = internalKvCoder.decode(inStream);
+      return TimestampedValue.of(kv.getValue(), Instant.ofEpochMilli(kv.getKey()));
+    }
+
+    @Override
+    public void verifyDeterministic() throws NonDeterministicException {
+      verifyDeterministic(
+          this, "TimestampedValueCoder requires a deterministic valueCoder", valueCoder);
```

Comment:
Just curious, why is this? It seems like we have no way of overwriting an individual element in the ordered list, so if it gets coded differently it wouldn't matter... It seems weird to be using non-deterministic coders anyway, but how would it affect correctness here?

Reply:
The coder here is based on the one defined at  https://github.com/apache/beam/blob/462c8334903c82917897cbe4c34d7aa8fe2bd375/sdks/java/core/src/main/java/org/apache/beam/sdk/values/TimestampedValue.java#L122.

Reply:
This is just part of the contract that coders must define, though I agree it's not something that would be likely to get leaked to somewhere that it matters...

Reply:
In this first version of implementation. I prefer keeping it there so it is consistent with the existing TimeStampedValueCoder defined in `TimestampedValue.java`. We can surely improve that later.